core/src/main/scala/org/http4s/parser/SimpleHeaders.scala

```diff
@@ -51,6 +51,21 @@ private[parser] trait SimpleHeaders {
         }
     }.parse
 
+  def ACCESS_CONTROL_ALLOW_METHODS(value: String): ParseResult[`Access-Control-Allow-Methods`] =
+    new Http4sHeaderParser[`Access-Control-Allow-Methods`](value) {
+      def entry =
+        rule {
+          oneOrMore(Token).separatedBy(ListSep) ~ EOL ~> { (ts: Seq[String]) =>
+            val ms = ts.map(
+              Method
+                .fromString(_)
+                .toOption
+                .getOrElse(sys.error("Impossible. Please file a bug report.")))
```

Comment:
We have a `bug` function in the internal package for this purpose.

Reply:
It doesn't look like a bug actually. It can happen if someone pass invalid/unsupported method in "Access-Control-Allow-Methods" header. Cannot it?

Reply:
If my assumption is correct, I'd suggest to use the `fail` or `failX` parboiled rule instead.

Reply:
All tokens are valid methods, even if they're not IANA-registered methods with a constant, so I think it is a bug if it fails.

In `Method`, the parser is in the companion object, so we can call the private constructor.  We use this pattern on a few model classes, which means we don't have these cases, but that pattern doesn't compose well, and we're left with ugly cases like this.  I'd like to rethink how we do that, but that's probably a separate ticket.

Reply:
Perhaps it makes sense to create a parser rule for Method which could produce a Method instance directly, something like `def METHOD: Rule1[Method]`...

Reply:
That would be useful, but there's still a design issue regarding where:

1. If the rule is somewhere in the parsers package, we need to expose an unsafe constructor outside `Method`.  That can be `private[http4s]`, but we have to be careful not to call `new Method(...)` in our internals.
2. We can put that rule in a trait in a `Parser` on the `Method` companion.  I started moving in this direction, but that contributes to the spaghetti of traits we have today.
3. We could have the rule like @amitrai48 wrote it in the parsers package, and only have that awful `bug` call once.
4. It would be a tiny bit more efficient to expose a `Method.unsafeFromString` and not wrap and unwrap an `Either` every time, which probably isn't that bad since we're just throwing anyway.  It just feels like defeat.

Maybe a combination of 3 and 4 is best.

Reply:
hi @rossabaker @satorg I am sorry I did not fully understand the above conversation. Ross mentioned that `All tokens are valid methods` does that mean `Access-Control-Allow-Methods: GET, xyz` is a valid header?

Reply:
I guess yes even though `xyz` is not an IANA-registered method.

Reply:
Yes, that's correct.  

I'm going to go deep into obscure spec lawyering here, and contradict one of my earlier claims.  Apologies for both.

I previously thought that `Access-Control-Allow-Methods: GET, *` is invalid, because it seems like wild cards are only intended as a lone value.  But on second thought, `*` is a token character, and thus is a technically valid method name!  It's sloppy that the spec imbues a valid value, however rare, with special semantics.  I don't see anything in the spec that tells us whether to treat that `*` as a wildcard or as a method.

The HTTP spec says that we should be able to join list headers with commas without changing the semantics.  That implies that:

```
Access-Control-Allow-Methods: GET
Access-Control-Allow-Methods: *
```

should be treated the same as:

```
Access-Control-Allow-Methods: GET, *
```

It would be highly weird that the wild card changes semantics when joined to another header.  I like to think of it as a semigroup.  I think nobody really intended `*` to be a method, and we should treat those as a wild card anywhere we see them.

In that case, instead of the sum type that we just introduced, maybe it's just a simple `NonEmptyList[Method]`.  And whoever processes this header can treat `Method("*")` as a wild card.